cloudsync/tests/fixtures/mock_provider.py

```diff
@@ -317,6 +333,36 @@ def _translate_event(self, pe: MockEvent, cursor) -> Event:
         retval = Event(standard_type, oid, path, None, not trashed, mtime, prior_oid, new_cursor=cursor)
         return retval
 
+    def _filter_event(self, event: Event) -> EventFilter:
+        # event filtering based on root path and event path
+
+        # Note: no filtering for oid-is-path providers for now
+        if not self._root_path or not self._filter_events or self.oid_is_path:
+            return EventFilter.PROCESS
+
+        state_path = self.sync_state.get_path(event.oid)
+        prior_subpath = self.is_subpath_of_root(state_path)
+        if not event.exists:
+            # delete - ignore if not in state, or in state but is not subpath of root
+            return EventFilter.PROCESS if prior_subpath else EventFilter.IGNORE
+
+        if event.path:
+            curr_subpath = self.is_subpath_of_root(event.path)
+            if curr_subpath and not prior_subpath:
+                # rename into root
+                log.debug("renamed into root: %s", event.path)
```

Comment:
Would this also be the case for new events?

Reply:
Yes, good catch, the log statement and comment should really say "create in or rename into root"

We can't differentiate between creates and renames without watching the entire filesystem -- the event has an oid and a current path, its a rename if we saw the oid before, but since we are ignoring events outside root we won't catch the case where something is created outside root and renamed into root. Hence the walk if its a directory -- a tradeoff for ignoring "outside root" events.

As a future enhancement we could track the oids for the entire filesystem but only act on (add to changeset, get_latest, etc.) events within root. That should allow us to differentiate renames and creates, and only do the walk on renames.